test/common/http/codec_client_test.cc

```diff
@@ -128,6 +131,82 @@ TEST_F(CodecClientTest, DisconnectBeforeHeaders) {
   connection_cb_->onEvent(Network::ConnectionEvent::RemoteClose);
 }
 
+TEST_F(CodecClientTest, IdleTimerWithNoActiveRequests) {
+  Http::StreamDecoder* inner_decoder;
+  NiceMock<Http::MockStreamEncoder> inner_encoder;
+  EXPECT_CALL(*codec_, newStream(_))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder) -> Http::StreamEncoder& {
+        inner_decoder = &decoder;
+        return inner_encoder;
+      }));
+
+  Http::MockStreamDecoder outer_decoder;
+  Http::StreamEncoder& request_encoder = client_->newStream(outer_decoder);
+  Http::MockStreamCallbacks callbacks;
+  request_encoder.getStream().addCallbacks(callbacks);
+  connection_cb_->onEvent(Network::ConnectionEvent::Connected);
+
+  Http::HeaderMapPtr response_headers{new TestHeaderMapImpl{{":status", "200"}}};
+  EXPECT_CALL(outer_decoder, decodeHeaders_(Pointee(Ref(*response_headers)), false));
+  inner_decoder->decodeHeaders(std::move(response_headers), false);
+
+  Buffer::OwnedImpl buffer("hello");
+  EXPECT_CALL(outer_decoder, decodeData(Ref(buffer), true));
+  inner_decoder->decodeData(buffer, true);
+  EXPECT_NE(client_->idleTimer(), nullptr);
+
+  // Close the client and validate idleTimer is reset
+  EXPECT_EQ(client_->numActiveRequests(), 0);
+  client_->close();
+  connection_cb_->onEvent(Network::ConnectionEvent::LocalClose);
```

Comment:
OK I see the issue, it looks like the CodecClient tests don't use the default connection mock handlers for raising events which is kind of a shame. We should add a TODO here to fix.